### PR TITLE
Feature/sagas

### LINKED ISSUE #26: Introduce Sagas to the sample app.
~~(And consider removing validation based on projections)~~

### PR DESCRIPTION
**Pull request by sluehr**

### CODE REVIEW

File: api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+        return organizationsReadService.getOrganizations().stream()
+                .map(dtoConverter::convert)
+                .collect(toList());
+    }
+
+    @PostMapping
+    @ResponseStatus(CREATED)
+    @ApiOperation("Register a new organization")
+    @AdminOnly
+    public UUID registerOrganization(User requestingUser, @RequestBody @Valid NewOrganizationRequest request) {
```
Comment (ywangd) [inline api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java]: Nit: Maybe make the method name consistent with the request name, i.e. `public UUID newOrganization(.., NewOrganizationRequest rquest)`. So it better differentiates from `OrganizationController#registerOrganization(RegisterOrganizationRequest)`.

File: api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+        return organizationsService.createRegisteredOrganization(requestingUser.getId(), request.getOrganizationName(),
```
Comment (ywangd) [inline api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java:64]: While organizaiton registration has a process to ensure email uniqueness, this request ends up to the persistent layer without any validation. Since the organizations table does not have any constraint on email column, this can result in corruption of email unqiueness guarantee.

File: api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+    public List<OrganizationResponse> getAllOrganizations() {
+        return organizationsReadService.getOrganizations().stream()
+                .map(dtoConverter::convert)
+                .collect(toList());
+    }
+
+    @PostMapping
+    @ResponseStatus(CREATED)
+    @ApiOperation("Register a new organization")
+    @AdminOnly
```
Comment (ywangd) [inline api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java:62]: Since this new controller is under an admin specific URL, `/api/admin/...`, it is better to configure the "admin only" access in Spring security configuration class, which does not have the risk of forgetting the annotation.

File: api/src/main/java/engineering/everest/lhotse/api/rest/requests/RegisterOrganizationRequest.java [hunk 4]
```diff
@@ -0,0 +1,27 @@
...
+        private String organizationName;
+        private String street;
+        private String city;
+        private String state;
+        private String country;
+        private String postalCode;
+        private String websiteUrl;
+        private String contactName;
+        private String contactPhoneNumber;
+        private String contactEmail;
```
Comment (ywangd) [inline api/src/main/java/engineering/everest/lhotse/api/rest/requests/RegisterOrganizationRequest.java:26]: I think the email field should be required too. Otherwise we have no way to contact user after initial registration.

File: api/src/test/java/engineering/everest/lhotse/api/config/TestApiConfig.java [hunk 5]
```diff
@@ -77,15 +77,18 @@ protected void configure(HttpSecurity httpSecurity) throws Exception {
...
+            );
 
             httpSecurity
                     .cors().and()
                     .csrf().disable()
                     .requestMatcher(requestMatcher)
                     .sessionManagement().sessionCreationPolicy(STATELESS).and()
                     .authorizeRequests()
-                    .antMatchers("/api/admin/**").hasRole(ADMIN.toString())
+                    .antMatchers("/admin/**").hasRole(ADMIN.toString())
```
Comment (ywangd) [inline api/src/test/java/engineering/everest/lhotse/api/config/TestApiConfig.java:91]: Why the change? The path of `AdminOrganizationsController` is `/api/admin/..`?

File: api/src/main/java/engineering/everest/lhotse/api/rest/controllers/OrganizationsController.java [hunk 6]
```diff
@@ -44,80 +45,69 @@
...
+        pendingRegistrationsService.registerOrganization(organizationId, userId, request.getOrganizationName(), request.getStreet(),
                 request.getCity(), request.getState(), request.getCountry(), request.getPostalCode(), request.getWebsiteUrl(),
-                request.getContactName(), request.getContactPhoneNumber(), request.getContactEmail());
+                request.getContactName(), request.getContactPhoneNumber(), request.getContactEmail(), request.getContactPassword());
+        return new OrganizationRegistrationResponse(organizationId, userId);
     }
 
-    @SuppressWarnings("PMD.AvoidDuplicateLiterals")
-    @GetMapping("/{organizationId}")
+    @GetMapping("/{organizationId}/register/{confirmationCode}")
```
Comment (ywangd) [inline api/src/main/java/engineering/everest/lhotse/api/rest/controllers/OrganizationsController.java]: I'd prefer POST here since domain data is altered by this call and it is not repeatable.

File: launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java [hunk 7]
```diff
@@ -58,25 +62,41 @@ void adminWillBeProvisioned() {
...
+        var pendingRegistration = pendingRegistrationsRepository.findByOrganizationId(newOrganizationId); // Confirmation code is "emailed", but we can cheat and pull it from the DB.
 
-        webTestClient.get().uri("/actuator/metrics/commandBus.successCounter")
-                .header("Authorization", "Bearer " + apiRestTestClient.getAccessToken())
-                .exchange()
-                .expectStatus().isEqualTo(OK);
+        apiRestTestClient.confirmOrganizationRegistration(newOrganizationId, pendingRegistration.getConfirmationCode(), OK);
+
+        apiRestTestClient.login("alice@example.com", "alicerocks");
+        Thread.sleep(15000L); // Allow event processing to catch up. CI is _ridiculously_ slow.
```
Comment (ywangd) [inline launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java]: What event processing is this waiting for? We use a single SEP (wrapped by SwitchingEP) and everything is synchronous. When `confirmOrganizationRegistration` returns, everything should be processed already? Also if the login call works before sleep, that means at least the user should be in place. Are you suspecting that the new organization may not have been created at this point? If so, I'd say we have a much bigger problem here because it is not just slow but a bug since it should *not* happen.

File: launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java [hunk 8]
```diff
@@ -58,25 +62,41 @@ void adminWillBeProvisioned() {
...
+        var newUserRequest = new NewUserRequest("bob@example.com", "bobalsorocks", "My name is Bob");
+        apiRestTestClient.createUser(newOrganizationId, newUserRequest, CREATED);
```
Comment (ywangd) [inline launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java:101]: This works for me locally. But I see CI failed on this line. Am I understanding this correctly: on each commit, two CI jobs will be triggered, one for the feature branch and one for the PR. They should be in theory identical. But somehow one fails and the other passes. It seems the one runs later always fail? I wonder whether the two runs could interfere with each other? This is just a wild guess. Their logs do suggest they were on different host. But these hosts may be able to talk to each other. If a HazelCast cluster somehow forms between them, it is likely to have strange behaviours.

Is there anyway that we can get detailed logs and stacktrace for the CI job?

File: organizations/src/main/java/engineering/everest/lhotse/organizations/domain/OrganizationAggregate.java [hunk 9]
```diff
@@ -27,77 +31,110 @@
...
+
+    private boolean isNameUpdated(UpdateOrganizationCommand command) {
+        return command.getOrganizationName() != null && !command.getOrganizationName().equals(organizationName);
+    }
+
+    private boolean areContactDetailsUpdated(UpdateOrganizationCommand command) {
+        return command.getWebsiteUrl() != null
                 || command.getContactName() != null
                 || command.getPhoneNumber() != null
                 || command.getEmailAddress() != null;
```
Comment (ywangd) [inline organizations/src/main/java/engineering/everest/lhotse/organizations/domain/OrganizationAggregate.java:131]: If we want to be consistent with how `isNameUpdated` is handled, these also need to check whether the content is actually changed.

File: organizations/src/main/java/engineering/everest/lhotse/organizations/domain/OrganizationAggregate.java [hunk 10]
```diff
@@ -27,77 +31,110 @@
...
+    }
 
-        Validate.isTrue(changesMade, "At least one organization field change must be requested");
+    private boolean isAddressUpdated(UpdateOrganizationCommand command) {
+        return command.getStreet() != null
+                || command.getCity() != null
+                || command.getState() != null
+                || command.getCountry() != null
+                || command.getPostalCode() != null;
```
Comment (ywangd) [inline organizations/src/main/java/engineering/everest/lhotse/organizations/domain/OrganizationAggregate.java:139]: Same as above.

File: pending-registrations-api/src/main/java/engineering/everest/lhotse/registrations/domain/commands/RegisterOrganizationCommand.java [hunk 11]
```diff
@@ -0,0 +1,37 @@
...
+
+@Data
+@AllArgsConstructor
+public class RegisterOrganizationCommand implements EmailAddressValidatableCommand, UserUniqueEmailValidatableCommand {
+
+    @TargetAggregateIdentifier
+    private UUID registrationConfirmationCode;
+    private UUID organizationId;
+    private UUID registeringUserId;
+    private String userEmailAddress;
```
Comment (ywangd) [inline pending-registrations-api/src/main/java/engineering/everest/lhotse/registrations/domain/commands/RegisterOrganizationCommand.java:21]: The `userEmailAddress` should be `@NotBlank` since we need it to communicate with user and `EmailAddressValidatableCommand` short-circuits for `null` value.

File: pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java [hunk 12]
```diff
@@ -0,0 +1,121 @@
...
+                    registeringUserDisplayName));
+        }
+    }
+
+    @SagaEventHandler(associationProperty = ASSOCIATION_PROPERTY)
+    public void on(UserCreatedForNewlyRegisteredOrganizationEvent event) {
+        commandGateway.send(new PromoteUserToOrganizationAdminCommand(organizationId, registeringUserId));
+        commandGateway.send(new CompleteOrganizationRegistrationCommand(event.getRegistrationConfirmationCode(),
+                organizationId, registeringUserId));
+    }
```
Comment (ywangd) [inline pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java]: Not sure whether we should think about this now since currently it is not an issue. For this handler to work, it assumes the same event is handled first by `UsersEventHandler`, where the user is created and persisted. So there is an ordering issue. Axon by default place Saga Handler Invoker after SimpleHandlerInvoker (for normal event handlers). So with the current code, this is guaranteed to run after the user is persisted. But we are relying on an Axon implementation detail, which could be brittle. In addition, what would happen when TEP is used? Something is needed here so we have the ordering guarantee regardlessly.

File: pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java [hunk 13]
```diff
@@ -0,0 +1,121 @@
...
+
+        commandGateway.send(new RecordSentOrganizationRegistrationEmailConfirmationCommand(event.getRegistrationConfirmationCode(),
+                event.getOrganizationId(), event.getRegisteringContactEmail(), event.getOrganizationName()));
+    }
+
+    @SagaEventHandler(associationProperty = ASSOCIATION_PROPERTY)
+    public void on(OrganizationRegistrationConfirmedEvent event) {
+        // The business may require that each email address be only available for registration once. In this case,
+        // the target aggregate identifier may be the user email address (possibly with an additional prefix).
+        if (usersReadService.hasUserWithEmail(registeringUserEmail)) {
```
Comment (ywangd) [inline pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java:91]: As we discussed before, this can possibly have a racing condition. Without a choking point, e.g. synchronisation, it is unlikely we can achieve total thread safety. Obviously, the downside of introducing synchronisation is performance. 

The tricky bit of this racing condition is that it could be totally random. There is no guarantee that during a replay the order would be the same, i.e. if Client_A succeeds and Client_B fails during normal application processing, when the system replays, it is possible that Client_A fails and Client_B succeeds!? This could be wreaking havoc.

File: users-persistence/src/main/java/engineering/everest/lhotse/users/persistence/PersistableUser.java [hunk 14]
```diff
@@ -39,7 +39,7 @@
 
     private boolean disabled;
     @ElementCollection(fetch = EAGER)
-    private Set<Role> roles =  EnumSet.noneOf(Role.class);
+    private Set<Role> roles = new HashSet<>();
```
Comment (ywangd) [inline users-persistence/src/main/java/engineering/everest/lhotse/users/persistence/PersistableUser.java]: Any reason for replacing `EnumSet` with `HashSet`?

File: pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java [hunk 15]
```diff
@@ -0,0 +1,121 @@
...
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.beans.factory.annotation.Qualifier;
+
+import java.util.UUID;
+
+@Saga
+@Revision("0")
+@Log4j2
+@NoArgsConstructor
+public class OrganizationRegistrationSaga {
```
Comment (ywangd) [inline pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java:31]: I still don't understand why Saga cannot be cached. Axon doc does have a brief [section](https://docs.axoniq.io/reference-guide/configuring-infrastructure-components/event-processing/saga-infrastructure#caching) talking about it. Could you please help me understand why it cannot be used in our case?

**Review by ywangd — COMMENTED** [review]
There are quite a lot of changes. I skipped most of the tests code, command and event classes so that I could keep the time under control and complete the first pass. I left some comments and questions.

File: api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java (hunk 1 above)
Comment (sluehr) [reply api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java]: :+1:

File: api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java (hunk 2 above)
Comment (sluehr) [reply api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java:64]: The thinking behind allowing multiple registrations for the same email address was to have a contrived example to demonstrate how such a situation can be handled by the saga and the command validation as it would be a business decision as to how to handle multiple pending registrations. 

I put a comment in the saga attempting to explain this.

Uniqueness is protected to the same level as the current command validation process as the final step of creating a user can still be rejected. (Cleaning up of the saga based on deadline expiry hasn't been implemented).

File: api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java (hunk 3 above)
Comment (sluehr) [reply api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java:62]: :+1:

File: api/src/main/java/engineering/everest/lhotse/api/rest/requests/RegisterOrganizationRequest.java (hunk 4 above)
Comment (sluehr) [reply api/src/main/java/engineering/everest/lhotse/api/rest/requests/RegisterOrganizationRequest.java:26]: :+1:

File: api/src/test/java/engineering/everest/lhotse/api/config/TestApiConfig.java (hunk 5 above)
Comment (sluehr) [reply api/src/test/java/engineering/everest/lhotse/api/config/TestApiConfig.java:91]: Yep

File: pending-registrations-api/src/main/java/engineering/everest/lhotse/registrations/domain/commands/RegisterOrganizationCommand.java (hunk 11 above)
Comment (sluehr) [reply pending-registrations-api/src/main/java/engineering/everest/lhotse/registrations/domain/commands/RegisterOrganizationCommand.java:21]: :+1:

File: pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java (hunk 13 above)
Comment (sluehr) [reply pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java:91]: If I understand the question correctly, your concern is that different commands will be dispatched during a replay? This won't happen as sagas do not handle events during replay.

File: pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java (hunk 15 above)
Comment (sluehr) [reply pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java:31]: It's a limitation of us relying on Hazelcast to assign aggregate ownership based on their IDs. In this saga, all commands dispatched to the organization aggregate are handled by the same node but the user aggregate commands are likely to be handled by different nodes.

When this happens, the second node loads the state of the saga from the database. The first node, however, doesn't seem to have any way of detecting that its version of the saga is out of date (from what I saw).

This happens to be safe for this saga as the saga state is not being changed as it's just reacting to events and dispatching commands but it seems like to cache it would be setting a trap for people.

File: pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java (hunk 12 above)
Comment (sluehr) [reply pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java]: Are you saying that the projections need to be updated before the user aggregate can receive the next event (the promotion to org admin)?

Events should be handled by the aggregates in the order that they appear in the event log. The command is being dispatched after the event has said to occur, so the order should be guaranteed.

If it's a question as to whether projections have had time to be updated then I think it's a similar situation. I'm not totally on top of the how the Axon internals work but even if aggregate state is ahead of projection state, the events should be projected in the same sequence for that aggregate.

Comment (sluehr) [reply pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java]: Ha! I think I was able to produce a race condition. I don't have time to address it right now, but I think we a solution may be to change the user creation command dispatch from the fire-and-forget send to the sendAndWait() call.


### Previous response:

Are you saying that the projections need to be updated before the user aggregate can receive the next event (the promotion to org admin)?

Events should be handled by the aggregates in the order that they appear in the event log. The command is being dispatched after the event has said to occur, so the order should be guaranteed.

If it's a question as to whether projections have had time to be updated then I think it's a similar situation. I'm not totally on top of the how the Axon internals work but even if aggregate state is ahead of projection state, the events should be projected in the same sequence for that aggregate.

File: users-persistence/src/main/java/engineering/everest/lhotse/users/persistence/PersistableUser.java (hunk 14 above)
Comment (sluehr) [reply users-persistence/src/main/java/engineering/everest/lhotse/users/persistence/PersistableUser.java]: A tired mind. I'll revert this.

File: api/src/main/java/engineering/everest/lhotse/api/rest/controllers/OrganizationsController.java (hunk 6 above)
Comment (sluehr) [reply api/src/main/java/engineering/everest/lhotse/api/rest/controllers/OrganizationsController.java]: Problem with a POST is that it's hard to do from an email client, which is where this link is most likely to be found :)

File: launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java (hunk 7 above)
Comment (sluehr) [reply launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java]: It was the next line that appears to fail - the second user on that org wasn't ready.

The sleep is a huge stink and obviously not the solution.

File: launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java (hunk 8 above)
Comment (sluehr) [reply launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java:101]: I was thinking the same, that Travis might have isolation issues.

I'll try and get the logs tomorrow.

File: organizations/src/main/java/engineering/everest/lhotse/organizations/domain/OrganizationAggregate.java (hunk 9 above)
Comment (sluehr) [reply organizations/src/main/java/engineering/everest/lhotse/organizations/domain/OrganizationAggregate.java:131]: :+1:

File: organizations/src/main/java/engineering/everest/lhotse/organizations/domain/OrganizationAggregate.java (hunk 10 above)
Comment (sluehr) [reply organizations/src/main/java/engineering/everest/lhotse/organizations/domain/OrganizationAggregate.java:139]: :+1:

File: launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java (hunk 8 above)
Comment (sluehr) [reply launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java:101]: Travis CI builds are definitely interfering with each other. The build just passed because one of the VMs was queued longer for creation.

Trying to get some log output to prove it.... multicast is probably enabled in their Google Cloud infrastructure.

Comment (sluehr) [reply launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java:101]: I've confirmed that the builds are isolated. However, there are two issues:
 - the saga is processed asynchronously, so the test does need to sleep for at least half a second for the subsequent events to complete processing
 - there seems to be a transaction boundary issue with the projection even handlers. This is what I'm focusing on now.

File: api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java (hunk 2 above)
Comment (ywangd) [reply api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java:64]: Ok I had some misunderstanding previously. This command creates organisation directly (as in directly into database), but does *not* create an user (unlike org registration). The email uniqueness is required for users, but not organizations, i.e. multiple organizations can share the same contact email. This makes sense.

However, this command has no validator, so it is also possible to create an organization with an email that is *not* associated with any user. I understand all these requirements are contrived. So they are somewhat arbitrary. But I'd prefer to have a validation to prevent this scenario from happening as it feels a bit strange.

File: api/src/main/java/engineering/everest/lhotse/api/rest/controllers/OrganizationsController.java (hunk 6 above)
Comment (ywangd) [reply api/src/main/java/engineering/everest/lhotse/api/rest/controllers/OrganizationsController.java]: Good one.

File: pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java (hunk 13 above)
Comment (ywangd) [reply pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java:91]: Ah OK I wasn't aware that Saga event handlers are *not* part of replay. It makes sense now.

File: pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java (hunk 15 above)
Comment (ywangd) [reply pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java:31]: Thanks for the explanation. This seems to be a major issue. Because Saga is meant to be cross aggregate and with Lhotse's setup, cross aggregate means multiple nodes. This basically makes Saga pretty useless as it cannot be stateful?

Is there any possible solution around this? Somehow route this event to the already loaded Saga instance in a way similar to distributed command is handled?

Comment (sluehr) [reply pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java:31]: It limits the ability to cache but it should still possible to hydrate the state from the DB every time an event is handled (the default)

File: launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java (hunk 8 above)
Comment (ywangd) [reply launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java:101]: I cannot reproduced with my 9 year old laptop. Did you fix it?

File: pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java (hunk 15 above)
Comment (ywangd) [reply pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java:31]: ok so its not a deal breaker. Phew ...

File: launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java (hunk 8 above)
Comment (sluehr) [reply launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java:101]: Nope, it's still reproducible. I haven't made much progress - too many other priorities.

File: api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java (hunk 2 above)
Comment (sluehr) [reply api/src/main/java/engineering/everest/lhotse/api/rest/controllers/AdminOrganizationsController.java:64]: The register command creates a "pending registration" rather than creating an organisation outright. In the current use case it's considered okay to have duplicate registrations as only the first confirmation for a given email address will create the actual organisation.

What hasn't been implemented is deadline management which would clean up orphaned registrations.

File: launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java (hunk 7 above)
Comment (sluehr) [reply launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java]: I'm going to move this discussion to a single thread now that we understand the issue that we were having.

File: launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java (hunk 8 above)
Comment (sluehr) [reply launcher/src/test/java/engineering/everest/lhotse/functionaltests/scenarios/ApplicationFunctionalTests.java:101]: I understand the problem better now. We have the same potential race condition for cross-aggregate event handlers even without sagas.

The solution is to do a sendAndWait() and to not have the saga responding to events that are being emitted while the method is still being invoked.

Normal command handling has an aggregate level lock that queues new commands. The saga was an example of applying commands to 3 aggregates at the same time, one of which happens to emit an event that is handled by an event handler not for that aggregate. So, for that one command only, it is necessary to sendAndWait().

The locking per aggregate makes sense when you consider that TEPs should be the way to go... design your aggregates so they don't interefere.

File: pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java (hunk 12 above)
Comment (sluehr) [reply pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java]: Closing this as we can discuss it in the single conversation (now that we understand it).

File: pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java [hunk 16]
```diff
@@ -83,37 +84,37 @@ public void on(OrganizationRegistrationReceivedEvent event) {
...
-                    registeringUserDisplayName));
+                    registeringUserDisplayName);
+            var promoteUserToOrganizationAdminCommand = new PromoteUserToOrganizationAdminCommand(organizationId, registeringUserId);
+            var completeRegistrationCommand = new CompleteOrganizationRegistrationCommand(event.getRegistrationConfirmationCode(),
+                    organizationId, registeringUserId);
+
+            commandGateway.send(registeredOrganizationCommand)
+                    .thenApply(fun -> commandGateway.sendAndWait(createUserCommand))
+                    .thenApply(fun -> commandGateway.sendAndWait(promoteUserToOrganizationAdminCommand))
+                    .thenApply(fun -> commandGateway.send(completeRegistrationCommand));
```
Comment (ywangd) [inline pending-registrations/src/main/java/engineering/everest/lhotse/registrations/domain/OrganizationRegistrationSaga.java]: The first command needs to be `sendAndWait` as well. Otherwise the whole event handling is still async. If there is any extra delay in Organization or User aggregate/handler (e.g. can be simulated adding `Thread.sleep(1000)` in `UsersEventHandler#on(UserPromotedToOrganizationAdminEvent event)`)  , the test would still fail.

I suggest change it to the followings:
```java
commandGateway.sendAndWait(registeredOrganizationCommand);
commandGateway.sendAndWait(createUserCommand);
commandGateway.sendAndWait(promoteUserToOrganizationAdminCommand);
commandGateway.send(completeRegistrationCommand);
```
Note the last command needs to be async, or we would run into deadlock.

**Comment by ywangd** [comment]
One more thing: It would be easier for incremental review if changes are not force pushed every time.
